### ISSUE TITLE
Add an option to disable enforcing of force_mask='700' for network file systems

### ISSUE THREAD

**Issue by TimSchneider42** [containers/storage#1839]
### Feature request description

Hi,

I run a small compute cluster for my research group in which we use rootless Podman. As file system for the user homes, we use BeeGFS (formerly FhGFS). BeeGFS seems to understand subuids, which is a crucial advantage over NFS for deploying rootless containers.

While running containers works fine, I am experiencing an issue with building images:
```bash
$ podman run -it ubuntu bash
WARN[0000] Network file system detected as backing store. Enforcing overlay option `force_mask="700"`.Add it to storage.conf to silence this warning
Resolved "ubuntu" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull docker.io/library/ubuntu:latest...
Getting image source signatures
Copying blob 57c139bbda7e done
Error: copying system image from manifest list: writing blob: adding layer with blob "sha256:57c139bbda7eb92a286d974aa8fef81acf1a8cbc742242619252c13b196ab499": processing tar file(lsetxattr /: operation not supported): exit status 1
```

I believe this issue is caused by the force_mask setting, which is applied because podman detects that BeeGFS is a network filesystem. If I trick Podman into believing that /home is a regular filesystem with an additional fuse-bindfs mount, it does not apply force_mask, and everything works fine, albeit at a performance penalty due to fuse-bindfs:

```bash
$ sudo mkdir /home-bindfs
$ sudo mount -t fuse.bindfs /home /home-bindfs
$ podman --root=/home-bindfs/$USER/.local/share/containers/storage run -it ubuntu bash
Resolved "ubuntu" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull docker.io/library/ubuntu:latest...
Getting image source signatures
Copying blob 57c139bbda7e done
Copying config fd1d8f58e8 done
Writing manifest to image destination
root@9521fdb6f3bf:/
```

I understand that network filesystems are currently not supported, but in this case, I feel it could be as easy as providing an option for disabling the force_mask setting for network filesystems. Unless the bindfs-trick does something else I am not aware of.

Thanks a lot in advance!

### Suggest potential solution

Adding an option for disabling the forced setting of force_mask="700" for network filesystems. Or adding an explicit exemption for BeeGFS, since it seems to work with Podman.

### Have you considered any alternatives?

A workaround is to trick Podman into believing the storage directory is not in a network filesystem with an additional fuse-bindfs mount as I described above. However, this has a serious impact on performance.

**Comment by giuseppe** [comment]
thanks, opened a PR: https://github.com/containers/storage/pull/1840

**Comment by TimSchneider42** [comment]
Thanks a lot for your help!

**Comment by jiridanek** [comment]
@TimSchneider42 are you using podman with this fix more extensively nowadays? How is it working for you? I tried to replicate a setup with rootless podman using a container storage directory on a BeeGFS mount, and while I can run various small images like the ubuntu you showed above, or fedora minimal, I see failures with running other images, such as

```
$ podman run --rm -it quay.io/sclorg/python-39-c9s:c9s
Trying to pull quay.io/sclorg/python-39-c9s:c9s...
Getting image source signatures
Copying blob 29f3d326b791 done   | 
Copying blob eff7947d9dec done   | 
Copying blob f2639e1c865c done   | 
Copying blob 81167cd56173 done   | 
Copying config d5906b34a8 done   | 
Writing manifest to image destination
ERRO[0123] Cleaning up container ea4ce5b577502cd17740dbec193b83934131c57adefb431785caa1bf547fccb5: unmounting container ea4ce5b577502cd17740dbec193b83934131c57adefb431785caa1bf547fccb5 storage: cleaning up container ea4ce5b577502cd17740dbec193b83934131c57adefb431785caa1bf547fccb5 storage: unmounting container ea4ce5b577502cd17740dbec193b83934131c57adefb431785caa1bf547fccb5 root filesystem: replacing mount point "/mnt/beegfs/containers/overlay/e4e8bd8327a1f510b63a6681d02b41a7ae2fe840421308f4fdd8eb754884953d/merged": file exists 
Error: creating temporary passwd file for container ea4ce5b577502cd17740dbec193b83934131c57adefb431785caa1bf547fccb5: container ea4ce5b577502cd17740dbec193b83934131c57adefb431785caa1bf547fccb5: open /mnt/beegfs/containers/overlay/e4e8bd8327a1f510b63a6681d02b41a7ae2fe840421308f4fdd8eb754884953d/merged/etc/group: device or resource busy
[cloud-user@jdanek-podman-builder-3 notebooks]$ podman run --rm -it quay.io/sclorg/python-39-c9s:c9s
ERRO[0000] Cleaning up container 6102cf27216181176102f3ed6daa6802314bc3e933bc9231b59e0447d93bceec: unmounting container 6102cf27216181176102f3ed6daa6802314bc3e933bc9231b59e0447d93bceec storage: cleaning up container 6102cf27216181176102f3ed6daa6802314bc3e933bc9231b59e0447d93bceec storage: unmounting container 6102cf27216181176102f3ed6daa6802314bc3e933bc9231b59e0447d93bceec root filesystem: replacing mount point "/mnt/beegfs/containers/overlay/adf52678131f9aa408f2253f32bf34371f39639f1608ddc7c4a131319c7fd895/merged": file exists 
Error: creating temporary passwd file for container 6102cf27216181176102f3ed6daa6802314bc3e933bc9231b59e0447d93bceec: container 6102cf27216181176102f3ed6daa6802314bc3e933bc9231b59e0447d93bceec: open /mnt/beegfs/containers/overlay/adf52678131f9aa408f2253f32bf34371f39639f1608ddc7c4a131319c7fd895/merged/etc/group: device or resource busy
[cloud-user@jdanek-podman-builder-3 notebooks]$ podman run --rm -it quay.io/sclorg/python-39-c9s:c9s bash
ERRO[0000] Cleaning up container 836accc0f7bc9c3a150d50d8db171602ee9c59b4c98a682dfde48bec2d6ba4ba: unmounting container 836accc0f7bc9c3a150d50d8db171602ee9c59b4c98a682dfde48bec2d6ba4ba storage: cleaning up container 836accc0f7bc9c3a150d50d8db171602ee9c59b4c98a682dfde48bec2d6ba4ba storage: unmounting container 836accc0f7bc9c3a150d50d8db171602ee9c59b4c98a682dfde48bec2d6ba4ba root filesystem: replacing mount point "/mnt/beegfs/containers/overlay/1eea2b56023f34e9a2c1df3c63c9e51dbadc6eb979c1c6254155df938e922649/merged": file exists 
Error: creating temporary passwd file for container 836accc0f7bc9c3a150d50d8db171602ee9c59b4c98a682dfde48bec2d6ba4ba: container 836accc0f7bc9c3a150d50d8db171602ee9c59b4c98a682dfde48bec2d6ba4ba: open /mnt/beegfs/containers/overlay/1eea2b56023f34e9a2c1df3c63c9e51dbadc6eb979c1c6254155df938e922649/merged/etc/group: device or resource busy
```

or with image builds

```

```

I installed BeeGFS by following the quickstart https://doc.beegfs.io/7.4.3/quick_start_guide/quick_start_guide.html#example-setup, using the RHEL 9 packages from https://www.beegfs.io/release/beegfs_7.4.3/dists/. I mistakenly used unsupported 9.4 and had to patch things up by installing older kernel (otherwise client kernel module would not build). I also set up the extended file attributes which podman seems to need by folowing https://doc.beegfs.io/7.3.4/advanced_topics/acl.html

I installed a supported kernel using

    sudo dnf install -y kernel-devel-5.14.0-362.24.1.el9_3.x86_64 kernel-5.14.0-362.24.1.el9_3.x86_64

To text extended file attributes, I did

    touch /mnt/beegfs/pepa
    python -c 'import os; os.setxattr("/mnt/beegfs/pepa", "user.lek", b"value", os.XATTR_CREATE)'

And to install podman, I tried first a centos koji build I found https://kojihub.stream.centos.org/koji/buildinfo?buildID=61248 and then the latest build on copr at https://copr.fedorainfracloud.org/coprs/rhcontainerbot/podman-next/. Both exhibited the unsatisfactory results from above.

I manged to make podman work when I switched from overlay to vfs, but that's glacially slow.

I don't feel like I have enough info collected for a competent issue ticket, so I'm just leaving a comment. Furthermore, the experience made me realize I don't want to do a network filesystem setup with podman. Instead of exposing the dist through BeeGFS for podmans on different machines to use, I should instead expose podman socket and use remote podman. That would work much better and there will be fewer low level problems.

**Comment by TimSchneider42** [comment]
Hi @jiridanek,

We have indeed experienced the same problems you described. It seems like `apt update` (or upgrade, don't remember) does not work, which makes it basically impossible to build images if the storage location is remote. The problem seems to stem from overlay-fs on top of BeeGFS, but I did not have time yet to properly isolate the issue, so I did not write an issue about it.

However, what works is pulling and importing images, which means that images can be built somewhere else and transferred into remote storage. It is a bit annoying but feasible in our case. What also works is using BeeGFS as an additional image store (via the `additionalimagestores` config option), which means that you can have shared images as long as the containers are local.

I am unsure if Podman actually supports multiple devices accessing a shared container storage location concurrently. I vaguely remember doing some tests with it (back when we were still on NFS) that did not go well, but I no longer remember the details.

So, to summarize, in our case, we have one machine that uses the BeeGFS storage as a main storage location for Podman. Unfortunately, it cannot build images, and running containers is also restricted due to the issues you mentioned. However, it can pull and import images into BeeGFS storage. All other machines in the network use the BeeGFS storage as an additional (read-only) image store. This works quite well except for the building problems on the main machine.

If you have any insights into this filesystem problem or any idea how to debug it, let me know.

Best,
Tim

**Comment by jiridanek** [comment]
I decided to try plain NFS instead of BeeGFS, where I encountered

* https://github.com/containers/podman/issues/22810

which is something I saw with BeeGFS too, but since I managed to reproduce this without any networked filesystem whatsoever, I reported it that way, hoping it will get more attention and will be easier to reproduce and fix.
